### ISSUE TITLE
search.js causes Content Security Policy violation with evaluated javascript

### ISSUE THREAD

**Issue by ArktekniK** [doxygen/doxygen#6381]
When hosting generated HTML documentation any on page that implements the content security protocol header with (even quite permissive values), the search functionality is broken. Specifically, this affects us when we are hosting API documentation on Jenkins CI, using the publishHTML api.

This is due to the fact that search.js attempts to use an unsafe eval when updating the search box with results. Although this instance of eval is implicit, the following page clarifies some of the security concerns. [See MDN here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#Do_not_ever_use_eval!)

Please see chrome error from console:
```
Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of 
script in the following Content Security Policy directive: "script-src 'self' 'unsafe-inline'".

SearchBox.OnSearchFieldChange @ search.js:204
```
You can see there that we have even relaxed the CSP header values, but it's not enough. To solve this issue, a server would need to send the `script-src unsafe-eval` value to the browser. Obviously this is a security concern, so I would like to propose a fix.

The offending code is here:
```js
if (searchValue != "") // non-empty search
{
    // set timer for search update
    this.keyTimeout = setTimeout(this.name + '.Search()', // implicit eval
                                 this.keyTimeoutLength);
}
```
I presume that the `name` is used to solve the issue whereby the function passed to `setTimeout` is executed on the `window` context, rather than the appropriate `this`. It is not necessary and there are better ways to solve that problem.

The following alternative will work. See [browser compatibility](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_objects/Function/bind#Browser_compatibility)
```
// set timer for search update
this.keyTimeout = setTimeout(this.Search.bind(this),  // Bind the correct context to the Search function
                              this.keyTimeoutLength);
```

Thanks!

**Comment by albert-github** [comment]
Robert,

Please create an issue for this in the issue tracker on github (
https://github.com/doxygen/doxygen/issues/new  doxygen recently moved bug
tracer from Bugzilla to Github!).


Albert

**Comment by ArktekniK** [comment]
Hi Albert,

Not sure I understand... I did file the bug on GitHub!

Rob

**Comment by albert-github** [comment]
Hi Rob,

Was a mistake from my side I thought it was coming from a doxygen forum, but a little bit later I saw that it was already from github (perfect), so I removed the comment again on github.
Sorry for the confusion.

Albert

**Comment by albert-github** [comment]
- for the fix it would be beneficial to have a pull request
- the `search.js` is automatically compiled into the doxygen executable (the base version is `templates/html/search.js`)
- the only way I see, due to the fact that the file resides in the subdirectory `search` of the `html` directory and the `HTML_EXTRA_FILES` are placed in the `html` directory, is to have in a postprocessing step copying the updated `search.js` to the `search` directory.

**Comment by bassdr** [comment]
`templates/html/navtree.js` and maybe others has the same issue.
https://github.com/doxygen/doxygen/blob/2972f6ddc491f615f814b1277ec715b6f707c419/templates/html/navtree.js#L438 

Should we create separate issue for it?

**Comment by albert-github** [comment]
@bassdr Problem is that your description is a bit vague (at least for me), do you see give the same problems? What is the relevance of the mentioned line in case of a problem?

Yest it would be good to have a more detailed explanation together with (some standard texts):
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem?  Please don't add external links as they might not be persistent.
- Please also specify the doxygen version used.

**Comment by bassdr** [comment]
@albert-github Here are more details: I get a JavaScript error:
```
Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of 
script in the following Content Security Policy
```
because the server I use to host the http is configured to reject script-src 'unsafe-eval' for security reasons. Any recent browser will then ignore the JavaScript and not execute it. End result is an empty treeview on the left when enabled in the doxyfile.

**Comment by albert-github** [comment]
I see in the `navtree.js` file the line you mentioned, but not jet the relevance in respect to the error message you get.
Is this the line that is indicated by the error / giving the error?

**Comment by bassdr** [comment]
@albert-github Yes, on chrome Version 78.0.3904.97 (64 bits), I have this exact error:

```
Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' 'unsafe-inline'".

    at HTMLScriptElement.<anonymous> (navtree.js:429)
```

And line 429 was, in doxygen 1.8.16 is:
``` javascript
navTreeSubIndices[i] = eval('NAVTREEINDEX'+i);
```
The `+i);` part is red underlined in the chrome debugger tab.

I also tried HEAD of master (d4a0825b1d00ead2ce441c3407d1983f80401353) with the same error. Microsoft Edge also refuses to execute the JavaScript. Did not try with Firefox.

Like I said, the problem is that the server is configured to deny `eval()` function by not setting `unsafe-eval` for `script-src` in the SCP. I am uploading the html to VisualSVN and I don't have control over how the SCP is configured on the server (as far as I understand it).

**Comment by albert-github** [comment]
I did some reading (https://stackoverflow.com/questions/30280370/how-does-content-security-policy-work) and a small test and it looks like the problems can be solved by setting:
```
<meta http-equiv="Content-Security-Policy" content="default-src 'self' 'unsafe-inline' 'unsafe-eval'">
```

Note: in search,js is another timeout setting (for closeWindow), even when correcting these 2 timout settings there are other issues as well.

@doxygen: should the proposed meta tag be added? as it is a security issue I'm not 100% sure whether or not we should do it.

**Comment by bassdr** [comment]
Hi @albert-github, I already tried it and no, it is overwritten by the server in my case. I already have a custom header with it and it does not help. unsafe-eval is globally disabled and un-overridable.

Anyhow, I still think it would be a good practice to add the http-equiv="Content-Security-Policy" in the header to let know what JavaScript of doxygen is expecting CSP-wise, in case the server let's you override it, or if the server does not configure CSP at all. But I think really the only option in my case would be to modify the javascript not to use eval at all. Maybe an option in the doxyfile like "ENABLE_HTML5" or something could switch to js versions without any eval. I am thinking about backward compatibility here...

**Comment by albert-github** [comment]
Pity that it doesn't work in your case.
I'm not an JavaScript expert so I don't know how to overcome the `eval` problem. The `setTimeout` problem can be solved like suggested in the original description.
Like I wrote `search.js` has more problems as I also got the message `SecurityError: Permission denied to access property "searchResults" on cross-origin object` and this points to line 203 in `search.js` i.e. `      else if (window.frames.MSearchResults.searchResults)`.
When looking further there might be more problems, that are now invisible in the console window.

**Comment by bassdr** [comment]
Thanks. I am no JS expert either. If I find a solution I'll open a PR. For now I disabled the tree view on the left nd the search altogether.

**Comment by jmirabel** [comment]
> Like I wrote `search.js` has more problems as I also got the message `SecurityError: Permission denied to access property "searchResults" on cross-origin object` and this points to line 203 in `search.js` i.e. ` else if (window.frames.MSearchResults.searchResults)`.

I encounter the same issue even when using local files.
Following [this](https://javascript.info/cross-window-communication), to by-pass the cross origin issue, it is possible to setup a communication between the windows as follow.
Instead of
```javascript
      else if (window.frames.MSearchResults.searchResults)
      {
        var elem = window.frames.MSearchResults.searchResults.NavNext(0);
        if (elem) elem.focus();
      }
```
by
```javascript
      else
      {
        window.frames.MSearchResults.postMessage("take_focus", "*");
      }
```
Then, in *search/all_n.html*, add this in the script tag
```javascript
window.addEventListener("message", function(event) {
  if (event.data == "take_focus") {
    var elem = searchResults.NavNext(0);
    if (elem) elem.focus();
  }
});
```

I checked this by modifying a generated doc. I can make a PR with this but I believe using the postMessage mechanism may be needed elsewhere.

**Comment by albert-github** [comment]
Code of pull request #7659 has been integrated in master on github (please don't close the issue as this will be done at the moment of an official release).

**Comment by jmirabel** [comment]
I believe this is only partially fixed since the tree view issue is not solved.

**Comment by albert-github** [comment]
@jmirabel will you be able to investigate this part as well?

**Comment by doxygen** [comment]
A possible solution for the use of `eval(x)` in the navtree script could be to use `window[x]` instead. Not sure if that has the same issues, but it seems to work the same for me (inspired by https://stackoverflow.com/a/724906).

**Comment by jmirabel** [comment]
> @jmirabel will you be able to investigate this part as well?

I am not a javascript expert and do this by trial and error with the help of a web search engine. Given that this issue is related to a server based documentation, It would be too time consuming to setup the failing scenario.

**Comment by albert-github** [comment]
Seen the fact that #9830 has been integrated into the master version of doxygen it would be good to know whether the "tree view issue" is now solved as well (problem mentioned in https://github.com/doxygen/doxygen/issues/6381#issuecomment-602216967)

@jmirabel is it possible for you to test this?

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.7.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
